File: functions/own_profile/get_my_groups.py
from firebase_admin import firestore
from flask import abort
from models.constants import Collections, GroupFields
from models.data_models import GroupsResponse, Group
from utils.logging_utils import get_logger


def get_my_groups(request) -> GroupsResponse:
    """
    Retrieves all groups where the current user is a member.
    
    This function queries the groups collection to find all groups that have the 
    authenticated user's ID in their members array. For each group, it retrieves 
    the basic information (groupId, name, icon, created_at).
    
    Args:
        request: The Flask request object containing:
                - user_id: The authenticated user's ID (attached by authentication middleware)
    
    Returns:
        A GroupsResponse containing:
        - A list of Group objects with basic information for each group
    """
    logger = get_logger(__name__)
    logger.info(f"Retrieving groups for user: {request.user_id}")

    # Get the authenticated user ID from the request
    current_user_id = request.user_id

    # Initialize Firestore client
    db = firestore.client()

    try:
        # Query for groups where the current user is a member
        groups_query = db.collection(Collections.GROUPS) \
            .where(GroupFields.MEMBERS, "array-contains", current_user_id) \
            .stream()

        logger.info(f"Querying groups for user: {current_user_id}")

        groups = []

        for doc in groups_query:
            group_id = doc.id
            group_data = doc.to_dict() or {}
            logger.info(f"Processing group: {group_id}")

            groups.append(Group(
                groupId=group_id,
                name=group_data.get(GroupFields.NAME, ""),
                icon=group_data.get(GroupFields.ICON, ""),
                members=group_data.get(GroupFields.MEMBERS, []),
                created_at=group_data.get(GroupFields.CREATED_AT, "")
            ))
            logger.info(f"Added group {group_id} to results")

        logger.info(f"Retrieved {len(groups)} groups for user: {current_user_id}")

        # Return the list of groups
        return GroupsResponse(
            groups=groups
        )
    except Exception as e:
        logger.error(f"Error retrieving groups for user {current_user_id}: {str(e)}", exc_info=True)
        # Use abort instead of returning empty response
        abort(500, "Internal server error")

File: functions/src/own_profile/get-my-groups.ts
import { Request, Response } from 'express';
import { getFirestore, QueryDocumentSnapshot, WhereFilterOp } from 'firebase-admin/firestore';
import { Collections, QueryOperators } from '../models/constants.js';
import { Group, GroupsResponse } from '../models/data-models.js';
import { GroupDoc, groupConverter, gf } from '../models/firestore/index.js';
import { getLogger } from '../utils/logging-utils.js';
import { formatTimestamp } from '../utils/timestamp-utils.js';

import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const logger = getLogger(path.basename(__filename));

/**
 * Retrieves all groups where the current user is a member.
 *
 * This function queries the group collection to find all groups that have the
 * authenticated user's ID in their member array. For each group, it retrieves
 * the basic information (groupId, name, icon, created_at).
 *
 * @param req - The Express request object containing:
 *              - userId: The authenticated user's ID (attached by authentication middleware)
 * @param res - The Express response object
 *
 * @returns A GroupsResponse containing:
 * - A list of Group objects with basic information for each group
 */
export const getMyGroups = async (req: Request, res: Response): Promise<void> => {
  const currentUserId = req.userId;
  logger.info(`Retrieving groups for user: ${currentUserId}`);

  const db = getFirestore();

  // Get all groups where the user is a member
  const groupsCol = db.collection(Collections.GROUPS).withConverter(groupConverter);
  const groupsQuery = groupsCol.where(gf('members'), QueryOperators.ARRAY_CONTAINS as WhereFilterOp, currentUserId);

  logger.info(`Querying groups for user: ${currentUserId}`);

  const groups: Group[] = [];

  // Process each group as it streams in
  for await (const doc of groupsQuery.stream()) {
    const groupDoc = doc as unknown as QueryDocumentSnapshot<GroupDoc>;
    const groupData = groupDoc.data();
    const groupId = groupDoc.id;
    logger.info(`Processing group: ${groupId}`);

    // Convert member_profiles from Record to array format
    const memberProfilesArray = Object.entries(groupData.member_profiles || {}).map(([userId, profile]) => ({
      user_id: userId,
      username: profile.username,
      name: profile.name,
      avatar: profile.avatar,
    }));

    groups.push({
      group_id: groupId,
      name: groupData.name || '',
      icon: groupData.icon || '',
      members: groupData.members || [],
      created_at: groupData.created_at ? formatTimestamp(groupData.created_at) : '',
      member_profiles: memberProfilesArray,
    });
    logger.info(`Added group ${groupId} to results`);
  }

  logger.info(`Retrieved ${groups.length} groups for user: ${currentUserId}`);

  // Return the list of groups
  const response: GroupsResponse = { groups };
  res.json(response);
};
